proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,678 @@
+``Exception`` class.
+
+
+Discussion
+----------
+
+The dynamically-typed open-world of exception types supported by Haskell is
+achieved through use of ``Typeable`` and the existentially-quantified
+``SomeException`` type (see [Marlow2006]_ for details). We
+extend this type to allow exceptions to be extended in the "product" sense,
+allowing users to decorate existing exception types with ad-hoc metadata
+(represented by the ``ExceptionContext`` type).
+
+The notion of ``ExceptionContext`` proposed here is taken from the generalized
+exception annotation machinery found in the ``annotated-exception`` `library
+<https://hackage.haskell.org/package/annotated-exception>`_, which demonstrated
+the utility of being able to attach ad-hoc contextual data to exceptions.
+By folding this notion into ``base``, we provide the community with a common
+means of capturing backtraces as well as application-specific metadata.
+
+GHC currently has four distinct mechanisms for capturing backtraces, each with
+its own backtrace representation:
+
+* ``HasCallStack``:
+   * Pros: Can be used on all platforms; provides precise backtraces
+   * Cons: Requires manual modification of the source program; runtime overhead
+* Cost-centre profiler (via ``GHC.Stack.CCS.getCurrentCCS``):
+   * Pros: Can be used on all platforms; fairly precise backtraces
+   * Requires profiled executable (``-prof``); runtime overhead; may require
+     manual ``SCC`` pragmas
+* DWARF debug information in conjunction with GHC's `built-in stack unwinder <https://www.haskell.org/ghc/blog/20200405-dwarf-3.html>`_:
+   * Pros: No runtime overhead; can trace through foreign code
+   * Cons: Highly platform-specific (currently only available on Linux); slow
+     backtrace collection; imprecise backtraces; large binary size overhead
+     (built with ``-g3``)
+* Info-table provenance (IPE) information (via ``GHC.Stack.CloneStack``):
+   * Pros: Can be used on all platforms; no runtime overhead
+   * Cons: Large binary size overhead; no visibility into foreign code; must be
+     built with ``-finfo-table-map``
+
+All of these backtrace mechanisms have their uses, offering a range of levels
+of detail, executable size, and runtime overhead. Given the complementary
+nature of these mechanisms, GHC should not dictate which of these mechanisms
+should be used to report exception backtraces. Consequently, we use the
+above-described context mechanism to allow backtraces from any of these
+mechanisms to be captured attached to exceptions.
+
+The fact that backtrace collection with some of these mechanisms can be
+rather expensive motivates two features of this proposal:
+
+* the ``NoBacktrace`` wrapper, allowing users to disable backtrace collection
+  at the ``throw``-site. This is sometimes necessary when exceptions are used
+  for non-exceptional control flow.
+
+* the ability to enable and disable individual exception mechanisms via
+  ``setEnabledBacktraceMechanisms``.
```

Comment:
This name is used here and below, but in the specification, above, you use `setBacktraceMechanismState` instead. I think they're not of the same type either.